barman/cloud_providers/google_cloud_storage.py

```diff
@@ -333,3 +337,282 @@ def _delete_objects_batch(self, paths):
+    implemented in Google Cloud Platform as documented at:
+
+        https://cloud.google.com/compute/docs/disks/create-snapshots
+    """
+
+    DEVICE_PREFIX = "/dev/disk/by-id/google-"
+
+    def __init__(self, project):
+        """
+        Imports the google cloud compute library and creates the clients necessary for
+        creating and managing snapshots.
+
+        :param str project: The name of the GCP project to which all resources related
+            to the snapshot backups belong.
+        """
+        if project is None:
+            raise TypeError("project cannot be None")
+        self.project = project
+
+        # The import of this module is deferred until this constructor so that it
+        # does not become a spurious dependency of the main cloud interface. Doing
+        # so would break backup to GCS for anyone unable to install
+        # google-cloud-compute (which includes anyone using python 2.7).
+        compute = import_google_cloud_compute()
+
+        self.client = compute.SnapshotsClient()
+        self.disks_client = compute.DisksClient()
+        self.instances_client = compute.InstancesClient()
+
+    def _get_instance_metadata(self, instance_name, zone):
+        """
+        Retrieve the metadata for the named instance in the specified zone.
+
+        :rtype: google.cloud.compute_v1.types.Instance
+        :return: An object representing the compute instance.
+        """
+        try:
+            return self.instances_client.get(
+                instance=instance_name,
+                zone=zone,
+                project=self.project,
+            )
+        except NotFound:
+            raise SnapshotBackupException(
+                "Cannot find instance with name %s in zone %s for project %s"
+                % (instance_name, zone, self.project)
+            )
+
+    def _get_disk_metadata(self, disk_name, zone):
+        """
+        Retrieve the metadata for the named disk in the specified zone.
+
+        :rtype: google.cloud.compute_v1.types.Disk
+        :return: An object representing the disk.
+        """
+        try:
+            return self.disks_client.get(
+                disk=disk_name, zone=zone, project=self.project
+            )
+        except NotFound:
+            raise SnapshotBackupException(
+                "Cannot find disk with name %s in zone %s for project %s"
+                % (disk_name, zone, self.project)
+            )
+
+    def take_snapshot(self, backup_info, disk_zone, disk_name):
+        """
+        Take a snapshot of a persistent disk in GCP.
+
+        :param barman.infofile.LocalBackupInfo backup_info: Backup information.
+        :param str disk_zone: The zone in which the disk resides.
+        :param str disk_name: The name of the source disk for the snapshot.
+        :rtype: str
+        :return: The name used to reference the snapshot with GCP.
+        """
+        snapshot_name = "%s-%s-%s" % (
```

Comment:
If the server and disk names are reasonably long then there is a risk of exceeding the maximum allowed length of a snapshot name in GCP (66 characters I think?).

Is it worth omitting the server name to give a little more breathing space?

Reply:
"The chain name must be 1-63 characters long and comply with RFC1035".
It is probably worth ensuring it does not exceed 63 characters. Dropping vm instance name is an acceptable solution.

Reply:
`backup_info.server_name` is going to be the name of the server as known by Barman rather than the instance name, but I am still thinking it's ok to drop it in return for more available characters in the disk name.

The only really important thing is that the snapshot name is unique for a given backup/disk combination and that's going to be true unless someone has attached the same disk to multiple servers (which hopefully isn't even possible?).

Reply:
This ended up being disk name + backup ID because snapshot names cannot start with numbers.